docs/design_notes/lift.md

```diff
@@ -0,0 +1,338 @@
+"""==>
+{
+    mlp: {
+        hidden: {
+            bias: (3, 4),
+            kernel: (3, 4, 4),
+        },
+        out: {
+            bias: (3, 1),
+            kernel: (3, 4, 1),
+        },
+    },
+}
+"""
+```
+
+Here `MLP(parent=None)` creates a detached instance of `MLP`.
+This avoids reserving a name for the submodule inside the current module.
+Although not strictly necessary, this also ensures we cannot accidentaly use the MLP instance in a stateful way and we are forced to use it through either `.init` or `.apply`.
+
+This example is still relatively concise but it already takes a few extra "bookkeeping" statements to make it work.
+However, this implementation has a number of limitations:
+1. During initialization we call the submodule twice through `init_fn` & `apply_fn`. If the submodule used the same trick to do
+   functional tranformation we will end up executing a lot of code as the number of module calls grows like 2^d where d is the number of
+   nested function transformations.
+2. The implemenation assumes the submodule only requires the parameter RNG sequence.
+3. The implementation assumes we init parameters once and we don't create any state variables either during init or apply.
+
+Point 3 in particular makes manual functionalization cumbersome. Feel free to try and extend the above example with a `nn.BatchNorm`
+layer in the `MLP` module. This should cause a number of complexites further growing the implementation. Like storing the updated
+batch stats after apply and making sure state isn't mutable inside vmap when it's immutable outside (eval mode).
+
+
+We call the process of transforming a stateful Module into a pure function "functionalization".
+By temporarily turning a stateful Module into a function we make it compatible with JAX's functional transformations.
+
+## Lifting
+
+
+Flax provides an alternative for manual functionalization which we call lifted transformation.
+Lifted transformations are defined in `flax.core.lift`.
+All the lifted JAX transformations are defined with a single generic lifting API called `pack`.
+
+A number of decision decisions had to be made in order to define `pack`. The implementation
+of `pack` controls how variables and rngs are lifted and how fine-grained the user control is.
+It must also decide whether lifting decisions are made during variable definition or at the transformation.
+
+
+### Lifting granularity
+
+
+With the linen API, users can define arbitrary variable collections and PRNG sequences.
+Each variable in a collection is lifted in the same way.
+
+Collections are typically given a semantically meaningful name like "params" or "batch_stats" rather than a general purpose name like "state".
+Because collections carry semantic meaning we can decide at the transformation level how each collection should be lifted.
+For example, we might decide to share parameters if we want to vectorize add a batch dimension to a model. 
```

Comment:
Is it "vectorize a batch dimension" without "add"? Or, is it "if we want to add a vectorized batch dimension to a model"

Another suggestion:

```suggestion
For example, we may decide to share parameters if we want to vectorize add a batch dimension to a model. 
```

(might < may/could < will)